rusk/src/lib/lib.rs

```diff
@@ -179,11 +180,12 @@ impl Rusk {
             }
         }
 
-        reward_and_update_root(
+        reward_slash_and_update_root(
```

Comment:
I am still unsure if the concept of slashing would be always fair. The more nodes are joining, the higher might be the network latency, thus we may encounter situations where the majority of committee members are running higher iteration than the block generator itself. In such cases, the generator might face unfair slashing.

PS. Haven't yet reproduced the scenario, the above statement is more a hunch than an established fact.

Reply:
I agree that we do need to ensure a generator is not slashed unfairly, just because of generalized network issues (I think we assume here that a network issue local to the generator's node is considered its fault and thus punished).

With respect to the changes in the latency due to joining nodes:
1. The Adaptive Timeout, if correctly designed, should allow nodes to cope with a normal variability in the network latency (assuming there are not hundreds of new nodes suddenly joining at once); if this is not enough, we can think about having provisioners better connected to each other than normal nodes (although I don't like the idea of distinguishing between provisioner and non-provisioner nodes at the network layer).

2. Timeout/"MissingCandidate" votes should be numerous enough to exclude a localized network problem. Currently we slash with a NilQuorum (1/3+1), but we could, if necessary, require a full quorum of "Timeout/Missing Candidate" votes instead. However, note that, in both cases, generalized issues at the network level (i.e. for many nodes in the network) mean that not only the Candidate message will be delayed but also the Validation messages, making it harder to reach a Quorum of "Timeout" votes.

Either way, I agree with your concern, and we should ensure slashing is fair before moving to Mainnet